### PR TITLE
On E0277 be clearer about implicit `Sized` bounds on type params and assoc types

### LINKED ISSUE #120178: Wonky diagnostics on unsatisfied implicit `Sized` bounds
### Code

```Rust
fn main() {
    let _ = std::mem::size_of::<[i32]>();
}
```

### Current output

```Shell
312 | pub const fn size_of<T>() -> usize {
    |                      ^ required by this bound in `size_of`
```

The "**this bound**" doesn't actually point to a bound.

### Desired output

Instead of "required by this bound"/"unsatisfied trait bound introduced here" we should say "note: `T` has an implicit bound of `Sized`"

### Rationale and extra context

_No response_

### Other cases

```rust
trait Foo {
    const VAL: usize;
}

impl<T> Foo for T {
    const VAL: usize = std::mem::size_of::<T>();
}

fn main() {
    println!("{}", <[i32] as Foo>::VAL);
}
```

Produces:

```Shell
5  | impl<T> Foo for T {
   |      -  ^^^     ^
   |      |
   |      unsatisfied trait bound introduced here
```

### Rust Version

```Shell
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0277]: the size for values of type `[i32]` cannot be known at compilation time
   --> f100.rs:2:33
    |
2   |     let _ = std::mem::size_of::<[i32]>();
    |                                 ^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[i32]`
note: required by an implicit `Sized` bound in `std::mem::size_of`
   --> /home/gh-estebank/rust/library/core/src/mem/mod.rs:312:22
    |
312 | pub const fn size_of<T>() -> usize {
    |                      ^ required by the implicit `Sized` requirement on this type parameter in `size_of`
```

Fix #120178.

### CODE REVIEW

**Comment by estebank** [comment]
This change *as is* affects incremental compilation, making the change `fn foo<T>() {}` to `fn foo<T: Sized>() {}` now be picked up as a change. Haven't looked at the incr compilation machinery in a while, so I'm not sure how long it might take me to fix that.

**Comment by fmease** [comment]
I will review this tomorrow together with your other PR :)

r? fmease

File: compiler/rustc_hir_analysis/src/astconv/bounds.rs [hunk 1]
```diff
@@ -82,8 +90,8 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
             );
         }
 
-        if seen_sized_unbound || seen_negative_sized_bound {
-            // There was in fact a `?Sized` or `!Sized` bound;
+        if seen_sized_unbound || seen_negative_sized_bound || seen_positive_sized_bound {
+            // There was in fact a `?Sized`, `!Sized` or `Sized` bound;
             // we don't need to do anything.
         } else if sized_def_id.is_some() {
             // There was no `?Sized` or `!Sized` bound;
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/astconv/bounds.rs:97]: I guess this could be updated, too. I know those comments are super repetitive and obvious. Proceed however you prefer.

Comment (estebank) [reply compiler/rustc_hir_analysis/src/astconv/bounds.rs]: Done

File: tests/ui/associated-types/issue-20005.stderr [hunk 2]
```diff
@@ -4,11 +4,11 @@ error[E0277]: the size for values of type `Self` cannot be known at compilation
 LL |     ) -> <Dst as From<Self>>::Result where Dst: From<Self> {
    |                                                 ^^^^^^^^^^ doesn't have a size known at compile-time
    |
-note: required by a bound in `From`
+note: required by an implicit `Sized` bound in `From`
   --> $DIR/issue-20005.rs:1:12
    |
 LL | trait From<Src> {
-   |            ^^^ required by this bound in `From`
+   |            ^^^ required by the implicit `Sized` requirement on this bound in `From`
```
Comment (fmease) [inline tests/ui/associated-types/issue-20005.stderr:11]: Could you rephrase this slightly? “requirement on this bound” isn't correct since `Src` isn't a bound but a generic parameter and a bound can't “have” a requirement (it *is* a requirement) and there can't be anything “on” a bound. Of course, this is preexisting but now it's slightly worse. I would go with something like this maybe:

```suggestion
   |            ^^^ required by the implicit `Sized` requirement on this type parameter in `From`
```

Comment (estebank) [reply tests/ui/associated-types/issue-20005.stderr]: Done.

**Review by fmease — APPROVED** [review]
Thanks a lot!
